crates/sparrow-instructions/src/evaluators/aggregation/first_evaluator.rs

```diff
@@ -0,0 +1,410 @@
+        let mut take_output_builder = UInt32Array::builder(input.len());
+        for input_index in 0..input.len() {
+            let entity_index = key_indices.value(input_index) as usize;
+            if token.value_is_null(entity_index) && input.is_valid(input_index) {
+                // If the `take_new_state[entity_index]` is greater than the length, that
+                // means it has been set already, so we should not overwrite it.
+                //
+                // Note that we only check if the token was previously null.
+                let not_taken = take_new_state[entity_index] < take_new_state.len() as u32;
+                if not_taken {
+                    take_new_state[entity_index] = (input_index + take_new_state.len()) as u32;
+                }
+            };
+
+            take_output_builder.append_value(take_new_state[entity_index])
+        }
+
+        // Gather the output, using the previous state and the new input
+        let output =
+            sparrow_arrow::concat_take(&token.accum, input, &take_output_builder.finish())?;
+
+        // Update the state token with the new state
+        let take_new_state = PrimitiveArray::from_iter_values(take_new_state);
+        let new_state = sparrow_arrow::concat_take(&token.accum, input, &take_new_state)?;
+        token.set_state(new_state);
+
+        Ok(output)
+    }
+
+    /// Returns the existing value for an entity if it exists, or a new value from the
+    /// input if it exists, or null if neither.
+    ///
+    /// Takes advantage of the `take` and `concat` kernels to avoid having to type the
+    /// evaluator, keeping everything as ArrayRefs.
+    ///
+    /// The output is taken from the concatenated batch of the old state and the new input.
+    /// If the old state's value is null, then the take index for that entity is the length
+    /// of the old state plus the current index (i.e. the index into the new input).
+    /// If not, then we keep the take index as the old state's index.
+    fn aggregate_since(
+        token: &mut ArrayRefAccumToken,
+        key_capacity: usize,
+        key_indices: &UInt32Array,
+        input: &ArrayRef,
+        ticks: &BooleanArray,
+    ) -> anyhow::Result<ArrayRef> {
+        Self::ensure_entity_capacity(token, key_capacity)?;
+
+        let mut take_new_state: Vec<Option<u32>> = (0..token.accum.len())
+            .map(|index| {
+                if token.accum.is_valid(index) {
+                    Some(index as u32)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        let mut take_output_builder = UInt32Array::builder(input.len());
+        for input_index in 0..input.len() {
+            let entity_index = key_indices.value(input_index) as usize;
+            if input.is_valid(input_index) && take_new_state[entity_index].is_none() {
+                take_new_state[entity_index] = Some((input_index + take_new_state.len()) as u32);
```

Comment:
I think we need to follow the same pattern as the above aggregate loop, with the `not_taken`, right?

Reply:
I don't think so -- `take_new_state` will only be `none` if there isn't a previous value, so it does the equivalent job.